core/src/main/java/org/elasticsearch/percolator/TopMatchingPercolatorType.java

```diff
@@ -0,0 +1,187 @@
+ */
+
+package org.elasticsearch.percolator;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.ReaderUtil;
+import org.apache.lucene.search.*;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.action.percolate.PercolateResponse;
+import org.elasticsearch.action.percolate.PercolateShardResponse;
+import org.elasticsearch.common.HasContextAndHeaders;
+import org.elasticsearch.common.bytes.BytesArray;
+import org.elasticsearch.common.text.BytesText;
+import org.elasticsearch.common.text.StringText;
+import org.elasticsearch.common.text.Text;
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.index.fieldvisitor.SingleFieldsVisitor;
+import org.elasticsearch.index.mapper.internal.UidFieldMapper;
+import org.elasticsearch.index.percolator.PercolatorQueriesRegistry;
+import org.elasticsearch.index.query.ParsedQuery;
+import org.elasticsearch.script.ScriptService;
+import org.elasticsearch.search.aggregations.InternalAggregations;
+import org.elasticsearch.search.highlight.HighlightField;
+import org.elasticsearch.search.highlight.HighlightPhase;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+class TopMatchingPercolatorType extends PercolatorType<TopScoreDocCollector> {
+
+    private final HighlightPhase highlightPhase;
+
+    TopMatchingPercolatorType(BigArrays bigArrays, ScriptService scriptService, HighlightPhase highlightPhase) {
+        super(bigArrays, scriptService);
+        this.highlightPhase = highlightPhase;
+    }
+
+    @Override
+    byte id() {
+        return 0x06;
+    }
+
+    @Override
+    PercolatorService.ReduceResult reduce(List<PercolateShardResponse> shardResults, HasContextAndHeaders headersContext) {
+        long foundMatches = 0;
+        int nonEmptyResponses = 0;
+        int firstNonEmptyIndex = 0;
+        for (int i = 0; i < shardResults.size(); i++) {
+            PercolateShardResponse response = shardResults.get(i);
+            foundMatches += response.count();
+            if (response.matches().length != 0) {
+                if (firstNonEmptyIndex == 0) {
+                    firstNonEmptyIndex = i;
+                }
+                nonEmptyResponses++;
+            }
+        }
+
+        int requestedSize = shardResults.get(0).requestedSize();
+
+        // Use a custom impl of AbstractBigArray for Object[]?
+        List<PercolateResponse.Match> finalMatches = new ArrayList<>(requestedSize);
+        if (nonEmptyResponses == 1) {
+            PercolateShardResponse response = shardResults.get(firstNonEmptyIndex);
+            Text index = new StringText(response.getIndex());
+            for (int i = 0; i < response.matches().length; i++) {
+                float score = response.scores().length == 0 ? Float.NaN : response.scores()[i];
+                Text match = new BytesText(new BytesArray(response.matches()[i]));
+                if (!response.hls().isEmpty()) {
```

Comment:
Yes... let me see what can be done here, I used move the already existing code to its own class.